File: src/main/java/com/tjh/riskfactor/repo/UserRepository.java
package com.tjh.riskfactor.repo;

import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.stereotype.Repository;

import com.tjh.riskfactor.entity.User;

import java.util.Optional;

@Repository
public interface UserRepository extends JpaRepository<User, Integer> {

    Optional<User> findByUsername(String username);

    boolean existsByUsername(String username);

    @Query(nativeQuery = true, value =
        "select count(*) from users a, users u where " +
        "a.id = :adminId and u.id = :userId and a.group_id = u.group_id and a.is_admin = true"
    )
    boolean isManaging(Integer adminId, Integer userId);

    @Query("select u.id from User u where u.username = :username")
    Optional<Integer> findIdByUsername(String username);

    @Query("select g.name from User u inner join u.group g on u.id = :uid")
    Optional<String> findGroupName(Integer uid);

    @Query(nativeQuery = true, value =
        "select u.group_id from users u where u.id = :uid and u.is_admin = true"
    )
    Optional<Integer> findManagedGroupId(Integer uid);

}

File: backend/src/main/java/com/tjh/riskfactor/repo/UserRepository.kt
package com.tjh.riskfactor.repo

import org.springframework.stereotype.Repository
import org.springframework.data.jpa.repository.Query
import org.springframework.data.jpa.repository.JpaRepository

import com.tjh.riskfactor.entity.User

@Repository
interface UserRepository: JpaRepository<User, Int> {

    fun findByUsername(username: String): User?

    fun existsByUsername(username: String): Boolean

    @Query(nativeQuery = true, value =
        "select count(*) from users a, users u where " +
            "a.id = :adminId and u.id = :userId and a.group_id = u.group_id and a.is_admin = true"
    )
    fun isManaging(adminId: Int, userId: Int): Boolean

    @Query("select u.id from User u where u.username = :username")
    fun findIdByUsername(username: String): Int?

    @Query("select g.name from User u inner join u.group g on u.id = :uid")
    fun findGroupName(uid: Int): String?

    @Query(nativeQuery = true, value =
        "select u.group_id from users u where u.id = :uid and u.is_admin = true"
    )
    fun findManagedGroupId(uid: Int): Int?

}
